flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobVertexScaler.java

```diff
@@ -383,23 +403,68 @@ protected static int scale(
         // Apply min/max parallelism
         newParallelism = Math.min(Math.max(parallelismLowerLimit, newParallelism), upperBound);
 
-        var adjustByMaxParallelism =
-                inputShipStrategies.isEmpty() || inputShipStrategies.contains(HASH);
-        if (!adjustByMaxParallelism) {
+        var adjustByMaxParallelismOrPartitions =
+                numSourcePartitions > 0 || inputShipStrategies.contains(HASH);
+        if (!adjustByMaxParallelismOrPartitions) {
             return newParallelism;
         }
 
-        // When the shuffle type of vertex inputs contains keyBy or vertex is a source, we try to
-        // adjust the parallelism such that it divides the maxParallelism without a remainder
-        // => data is evenly spread across subtasks
-        for (int p = newParallelism; p <= maxParallelism / 2 && p <= upperBound; p++) {
-            if (maxParallelism % p == 0) {
+        final int numKeyGroupsOrPartitions;
+        final int upperBoundForAlignment;
+        if (numSourcePartitions <= 0) {
+            numKeyGroupsOrPartitions = maxParallelism;
+            upperBoundForAlignment =
+                    Math.min(
+                            // Optimize the case where newParallelism <= maxParallelism / 2
+                            newParallelism > maxParallelism / 2
+                                    ? maxParallelism
+                                    : maxParallelism / 2,
+                            upperBound);
+        } else {
+            numKeyGroupsOrPartitions = numSourcePartitions;
+            upperBoundForAlignment = Math.min(numSourcePartitions, upperBound);
+        }
+
+        // When the shuffle type of vertex inputs contains keyBy or vertex is a source,
+        // we try to adjust the parallelism such that it divides
+        // the adjustableMaxParallelism without a remainder => data is evenly spread across subtasks
+        for (int p = newParallelism; p <= upperBoundForAlignment; p++) {
+            if (numKeyGroupsOrPartitions % p == 0) {
```

Comment:
About this  comment https://github.com/apache/flink-kubernetes-operator/pull/879#discussion_r1764419949, I'm thinking whether the following change is more reasonable?

Note: `numKeyGroupsOrPartitions / p` means how many source partitions or key groups every subtask consume.

```suggestion
            if (numKeyGroupsOrPartitions % p == 0 || numKeyGroupsOrPartitions / p < numKeyGroupsOrPartitions / newParallelism) {
```

For example: maxParallelism is 200, and new parallelism is 60. (Some subtasks consume 4 keyGroups, the rest of subtask consume 3 keyGroups)

- The final parallelism is 100 based on the main branch code due to we only return p when `maxParallelism % p == 0`.
- But I think 67 is more reasonable here. (One subtask consumes 2 key groups. The remaining 66 subtasks, each subtask consumes 3 key groups.)


@mxm @gyfora , WDYT? 

Also, it's a bit beyond the scope of this PR. I could file a separate PR if you think it makes sense. Of course, it's acceptable to be done at this PR.

Reply:
I think that makes sense, but it makes the scaling more aggressive and less balanced. If we want to be more conservative, maybe 100 is ok in this scenario, where there is actually a divisor without a remainder. When there isn't, I think what you propose is way better than just using the initially provided parallelism.

In summary, I'm proposing to do a two-step process, similarly as for the partitions, where we first try to find a parallelism that divides the key groups without a remainder, and if that fails we do what you propose.

Reply:
Thanks for the review, I agree with this, We can introduce an additional parameter to enable a more aggressive strategy: https://github.com/apache/flink-kubernetes-operator/pull/879#discussion_r1764419949. But by default, a divisor of the number of partitions is still used to ensure balanced consumption.
WDYT？ @mxm @1996fanrui

Reply:
> I think that makes sense, but it makes the scaling more aggressive and less balanced. If we want to be more conservative, maybe 100 is ok in this scenario, where there is actually a divisor without a remainder. When there isn't, I think what you propose is way better than just using the initially provided parallelism.
> 
> In summary, I'm proposing to do a two-step process, similarly as for the partitions, where we first try to find a parallelism that divides the key groups without a remainder, and if that fails we do what you propose.


Thanks for the review, I agree with this, We can introduce an additional parameter to enable a more aggressive strategy: https://github.com/apache/flink-kubernetes-operator/pull/879#discussion_r1764419949. But by default, a divisor of the number of partitions is still used to ensure balanced consumption.
WDYT？ @mxm @1996fanrui

Reply:
Sounds make sense to me, let us choose a conservative solution in this PR first. 

We could improve it in the future if it's needed.

Reply:
Sounds make sense to me, we could introduce an `additional parameter to enable a more aggressive strategy`.

I wanna check with @huyuanfeng2018 and @mxm :  The strategy will work for both source partition and key group, right? As I understand, we could unify the strategy for these 2 cases.

Reply:
> I wanna check with @huyuanfeng2018 and @mxm : The strategy will work for both source partition and key group, right? As I understand, we could unify the strategy for these 2 cases.

I think we can unify strategies

Reply:
Sounds good!